diff --git a/src/main/java/neoflix/services/AuthService.java b/src/main/java/neoflix/services/AuthService.java
index d8a5487..8c83d07 100644
--- a/src/main/java/neoflix/services/AuthService.java
+++ b/src/main/java/neoflix/services/AuthService.java
@@ -5,6 +5,7 @@ import neoflix.AuthUtils;
 import neoflix.ValidationException;
 import org.neo4j.driver.Driver;
 import org.neo4j.driver.Values;
+import org.neo4j.driver.exceptions.ClientException;
 
 import java.util.List;
 import java.util.Map;
@@ -73,7 +74,15 @@ public class AuthService {
             // tag::return-register[]
             return userWithToken(user, token);
             // end::return-register[]
+        // tag::catch[]
+        } catch (ClientException e) {
+            // Handle unique constraints in the database
+            if (e.code().equals("Neo.ClientError.Schema.ConstraintValidationFailed")) {
+                throw new ValidationException("An account already exists with the email address", Map.of("email","Email address already taken"));
+            }
+            throw e;
         }
+        // end::catch[]
     }
     // end::register[]
 
